lib/ansible/template/__init__.py

```diff
@@ -152,10 +157,186 @@ def _count_newlines_from_end(in_str):
         while in_str[j] == '\n':
             j -= 1
         return i - 1 - j
+    except TypeError:
+        return 0
     except IndexError:
         # Uncommon cases: zero length string and string containing only newlines
         return i
 
+
+def ansible_j2_concat(invals):
+    '''
+    Jinja's native concat is just a list join. The list comes from result of
+    templating all the nodes. If the return type is expected to be
+    native, the list is artificial and we should return just the first
+    element.
+    '''
+    invals = [x for x in invals]
+    if isinstance(invals, list):
+        if len(invals) == 1:
+            # break out the single value
+            invals = invals[0]
+        elif len(invals) > 1:
+            # cast to unicode and join
+            invals = u''.join([u'%s' % x for x in invals])
+    return invals
```

Comment:
We could make this more efficient.  what types can be given here?  Since the original jinja2 code is u"".join(invals) it seems like the minimum assumptions we can make are that invals is an iterable and that all the elements can convert to a text type without unicode exceptions.  So maybe something like this:

``` python
# We can omit this block if invals is never a string type
if isinstance(invals, six.text_type):
    return invals
elif isinstance(invals, six.binary_type):
    return to_text(invals, errors='surrogate_or_strict')

try:
    if len(invals) == 1:
        # Break out the single value
        return invals[0]
except TypeError:
    # Iterable (by contract) but not a Sequence
    pass

return u''.join(invals)
```

* This keeps us from instantiating two lists.
* This should be faster even than the jinja2 concat if strings (not inside of a container) are allowed as invals.
* If we don't have to deal with strings, then we can get rid of that block and reduce the number of conditionals.

I can adapt that further if we need to make less assumptions (or if we can make more of them).